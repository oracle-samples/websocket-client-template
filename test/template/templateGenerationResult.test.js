/******************************************************************************
 *
 * Copyright (c) 2024 Oracle and/or its affiliates.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 *****************************************************************************/

/**
 * @jest-environment node
 */

const { readFile } = require('fs').promises;
const path = require('path');
const Generator = require('@asyncapi/generator');
const outputDir = path.resolve('test/temp/templateGenerationResult', Math.random().toString(36).substring(7));

describe('templateGenerationResult()', () => {
  jest.setTimeout(1000000);

  const params = {
    server: 'localhost'
  };

  const expectedFiles = [
    'client.js',
    'client.py'
  ];

  beforeAll(async() => {
    const generator = new Generator('./', outputDir, { forceWrite: true, templateParams: params });
    await generator.generateFromFile(path.resolve('test','streaming.yaml')); 
  });

  it('generated correct streaming client code', async () => {
    for (const fileName of expectedFiles) { 
      /* eslint-disable-next-line security/detect-non-literal-fs-filename -- Safe as no value holds user input */
      const file = await readFile(path.join(outputDir, fileName), 'utf8');
      expect(file).toMatchSnapshot();
    }
  });
});
