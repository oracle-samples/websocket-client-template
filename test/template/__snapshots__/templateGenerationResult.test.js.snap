// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`templateGenerationResult() generated correct streaming client code 1`] = `
"//////////////////////////////////////////////////////////////////////
//
// data streaming API - 1.0.0
// ws protocol: 
// admin:admin@localhost:9002 
// /services/v2/stream/stream1
//////////////////////////////////////////////////////////////////////
const WebSocket = require('ws')
const reader = require('readline-sync');
const fs = require('fs');

//////////////////////////////////////////////////////////////////////
//
// This client demonstrates the one-way websocket streaming use case
// -- It assumes only one channel in the server!
// -- It assumes only 'subscribe' oepration in the channel!
// -- It supports query parameters such as ?begin=now&format=json
//
//////////////////////////////////////////////////////////////////////

////////////////////////////////////////////////////////////////
//
// generic data processing with the websocket service,
// assume an array of json objects.
//
////////////////////////////////////////////////////////////////
const onRecords = (wsClient) => {
    wsClient.on('message', function message(data) {
        console.log('received some data:')
        const records = data.toString()
 
        const recordArray = eval(records);
        for (var i = 0; i < recordArray.length; i++) {
	    console.log(recordArray[i]);
            //data processing, implement user logic here. 
        }
        
    });
    wsClient.on('error', (err) => {
        console.log(err.toString());
    });
}

////////////////////////////////////////////////////////////////
//
// main entry point for the example client:
// asyncapi yaml definition is parsed to provide service
// access URL and a dedicated websocket connection is
// created to stream data from the service.
//
////////////////////////////////////////////////////////////////

const init = async () =>{
       
       
    let serviceURL = 'ws://admin:admin@localhost:9002/services/v2/stream/stream1';
      const queryParams = new Map([['begin','earliest']]);

    const queryParamSign    = "?";
    const queryParamDelimit = "&";
    let queryParamSignAdded = false;
    let size                = queryParams.length;

    for (const [key, value] of queryParams) {
      let paramValue = reader.question("Enter the value for query parameter "+key+"(default="+value+"):") || value;

      if (!queryParamSignAdded) {
        serviceURL += queryParamSign;
        queryParamSignAdded = true;
      }
      serviceURL += key + "=" + paramValue;
      size--;
      if (size) {
        serviceURL += queryParamDelimit;
      }
    }
  
    console.log(" ");
    console.log("Establishing websocket connection to: ");
    // uncomment below for debugging
    console.log(serviceURL); 
    console.log(" ");
 
    // establishing websocket connection to the service
    const wsClient = new WebSocket(serviceURL);
    
    console.log(" ");
    console.log("Start streaming data from the service ...");
    console.log(" ");

    // now start the client processing    
    onRecords (wsClient)
}

init()
"
`;    

exports[`templateGenerationResult() generated correct streaming client code 2`] = `
"#!/usr/bin/env python
###############################################################################
#
# data streaming API - 1.0.0
# ws protocol: 
# admin:admin@localhost:9002 
# /services/v2/stream/stream1
###############################################################################

import asyncio
import websockets
import json
import ssl
import pathlib
import getpass
import os

###############################################################################
#
# This client demonstrates the one-way websocket streaming use case
# -- It assumes only one channel in the server!
# -- It assumes only 'subscribe' oepration in the channel!
# -- It supports query parameters such as ?begin=now&format=json
#
###############################################################################

###############################################################################
#
# main entry point for the example client:
# asyncapi yaml definition is parsed to provide service
# access URL and a dedicated websocket connection is
# created to stream data from the service.
#
###############################################################################
async def onRecords():
    ## take user input
 
    # construct service URL
    serviceURL = 'ws://admin:admin@localhost:9002/services/v2/stream/stream1'
      
    queryParamsDefault = {}
    queryParamsDefault['begin'] = 'earliest'


    queryParamSign      = '?'
    queryParamDelimit   = '&'
    queryParamSignAdded = False
    size                = len(queryParamsDefault)

    for key in queryParamsDefault.keys():
      prompt = "Enter value for query parameter "+key+"(default="+queryParamsDefault[key]+"):"
      paramValue = input(prompt) or queryParamsDefault[key];

      if not queryParamSignAdded:
        serviceURL += queryParamSign
        queryParamSignAdded = True
      serviceURL += key + "=" + paramValue
      size=size-1
      if size:
        serviceURL += queryParamDelimit
     
    # uncomment below for debugging
    #print(serviceURL)    
 
    ## establishing websocket connection to the service
    async with websockets.connect(serviceURL) as websocket:
        print("start streaming data from the service")
        while True:
            data = await websocket.recv()
            records = json.loads(data)
            count = len(records)
            print(f"< received {count} records")
            for rec in records:
                #
                # implement client side data processing logic here
                #
                print(f"< {rec}")

asyncio.run(onRecords())
"
`;    
